src/v/strings/static_str.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include <fmt/format.h>
+
+#include <string_view>
+#include <type_traits>
+
+#pragma once
+
+/**
+ * @brief A wrapper around a compile-time constant string.
+ *
+ * A static_str wraps an compile-time constant string, and is intended to be
+ * used at runtime as a guarantee that a string has static lifetime. Such string
+ * have trivial and cheap copy operations, since we just copy pointers (or
+ * views) to the underlying string.
+ */
+class static_str {
+public:
+    /**
+     * @brief Create a static_str based on a compile-time literal.
+     *
+     * The use of consteval prevents passing anything other than compile-time
+     * values.
+     */
+    consteval static_str(const char* str) // NOLINT
+      : _data{str} {
+        // do a compile-time check that the string is zero-terminated
+        std::string_view{str};
+    }
+
+    constexpr bool operator==(const static_str& rhs) const {
+        if (!std::is_constant_evaluated() && _data == rhs._data) {
+            // fast path based on pointer comparison, which is a sufficient but
+            // not necessary condition for equality
+            return true;
+        }
+        return *this == static_cast<std::string_view>(rhs);
+    }
+
+    constexpr auto operator<=>(const static_str& rhs) const {
+        if (!std::is_constant_evaluated() && _data == rhs._data) {
```

Comment:
I don't think it's the main thing, I don't even use these operators in the use I'm introducing (handy for tests though), and I almost didn't put this fast path per YAGNI. The main thing is very efficient copy & single-register size.

Still, this optimization is handy in some cases: it often makes comparison very cheap O(1) in practice since either the strings are equal and this case is used (not guaranteed but every very in practice for statically linked binaries: strings are generally de-duped), or they are unequal but the comparison fails quickly in the first few characters.

BTW, the `!is_constant_evaluated` is needed because constexpr eval complains that comparing string literal addresses is unspecified (which is true, but this code should return the right result either way).